### PR TITLE
Keep global numbering for test cases with `count: X` (#271)

### PR DESCRIPTION
**Pull request by mzuenni**
not sure if this is the best solution...

### CODE REVIEW

File: bin/generate.py [hunk 1]
```diff
@@ -1275,6 +1275,20 @@ def add_known(obj):
         num_numbered_testcases = 0
         testcase_id = 0
 
+        def parse_count(yaml, warning=False):
+            if yaml is None or 'count' not in yaml or not isinstance(yaml['count'], int):
```
Comment (mpsijm) [inline bin/generate.py:1279]: Do we want to warn if `count` is given, but not an `int`?

File: bin/generate.py [hunk 2]
```diff
@@ -1291,44 +1305,39 @@ def count(yaml):
...
                 ts = []
                 for count_index in range(count):
-                    curname = name if count == 1 else name + f'-{count_index+1:0{len(str(count))}}'
+                    if count_index > 0:
+                        name = name_gen()
 
                     # If a list of testcases was passed and this one is not in it, skip it.
-                    if not self.process_testcase(parent.path / curname):
+                    if not self.process_testcase(parent.path / name):
                         continue
```
Comment (mpsijm) [inline bin/generate.py:1338]: I have the feeling that this does not really have a proper semantics for test cases with a `count:` :thinking:

File: bin/generate.py [hunk 3]
```diff
@@ -1363,24 +1372,34 @@ def parse(key, name, yaml, parent):
...
+                                    )
+
+                                child_name = next_testgroup_name
                             elif is_testcase(child_yaml):
-                                testcase_id += 1
-                                child_name = numbered_testcase_name(
-                                    child_key, testcase_id, num_numbered_testcases
-                                )
+
+                                def next_testcase_name():
```
Comment (mpsijm) [inline bin/generate.py:1386]: Perhaps we can move the `next_testgroup_name` and `next_testcase_name` outside of the loops, to make sure that Python does not re-create them with every iteration :slightly_smiling_face:

**Review by mpsijm — COMMENTED** [review]
I was thinking of a similar approach, it isn't as bad as I feared :smile:

File: bin/generate.py (hunk 2 above)
Comment (mzuenni) [reply bin/generate.py:1338]: testcases with count behave the same way as any other... i.e. have regular names?

File: bin/generate.py (hunk 3 above)
Comment (mzuenni) [reply bin/generate.py:1386]: we need to know child_key?

File: bin/generate.py (hunk 2 above)
Comment (mpsijm) [reply bin/generate.py:1338]: I never actually used `bt generate [testcases...]`, so I'm also not 100% sure what the syntax should be for the testcases that should be generated. Any syntax that I try, results in `bt generate` simply printing "Done" twice, with nothing happening, even though I make changes to the relevant test cases in `generators.yaml` :thinking:

Comment (mzuenni) [reply bin/generate.py:1338]: you should give it a path like `data/secret/blub.in` i guess?

File: bin/generate.py (hunk 1 above)
Comment (mzuenni) [reply bin/generate.py:1279]: this happens inside the TestcaseRule

**Comment by mzuenni** [comment]
Hm now count only works if the directory is numbered... i think we should still append the count_index?

**Review by mpsijm — APPROVED** [review]
Thanks for implementing this! :smile:

**Pull request merged**